Core: Increase compile targets for node & browsers

JReinhold opened the pull request:
<!-- If your PR is related to an issue, provide the number(s) above; if it resolves multiple issues, be sure to break them up (e.g. "closes #1000, closes #1001"). -->

<!--

Thank you for contributing to Storybook! Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/contribute

-->

## What I did

This PR updates the browser and Node.js targets for Storybook 9:

- Chrome 131
- Safari 18.3
- Edge 134
- Firefox 136
- Opera 117 ([Chromium 132](https://blogs.opera.com/desktop/2025/02/opera-117/))
- Node.js 20

This marks the start of a new strategy, where the browsers and versions we support will be based on usage data from Storybook docs visitors combined with generally accepted heuristics inspired by [browserslist's `default`](https://browsersl.ist/#q=defaults).

1. A follow-up PR will go into more detail about this, and add documentation for the support as well.
2. https://github.com/storybookjs/storybook/pull/31102 will add a "preview only" feature that mitigates the problem this narrower browser support can create, by allowing users to navigate directly from within the preview, which is and has always been under the user's full control when it comes to browser support and compile target.
3. Another potential follow-up to this would be to show a warning (notification) when an unsupported browser is detected, still TBD.

<!-- Briefly describe what your PR does -->

## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:

- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

1. Open a sandbox and see that it still works with the browser versions listed above.

<!-- Please include the steps to test your changes here. For example:

1. Run a sandbox for template, e.g. `yarn task --task sandbox --start-from auto --template react-vite/default-ts`
4. Open Storybook in your browser
5. Access X story

-->

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

  - `bug`: Internal changes that fixes incorrect behavior.
  - `maintenance`: User-facing maintenance tasks.
  - `dependencies`: Upgrading (sometimes downgrading) dependencies.
  - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
  - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
  - `documentation`: Documentation **only** changes. Will not show up in release changelog.
  - `feature request`: Introducing a new feature.
  - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
  - `other`: Changes that don't fit in the above categories.

   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

<!-- CANARY_RELEASE_SECTION -->

<!-- BENCHMARK_SECTION -->
<!-- BENCHMARK_SECTION -->


<!-- greptile_comment -->

## Greptile Summary

Updates browser and Node.js targets for Storybook 9, implementing stricter compatibility requirements based on usage data from Storybook docs visitors.

- Sets new minimum browser versions in `code/core/src/builder-manager/index.ts`: Chrome 130, Safari 18.1, Edge 132, Firefox 134
- Updates Node.js target to version 20 in `code/core/scripts/prep.ts`
- Adds ESLint browser compatibility checking via 'compat' plugin in `code/.eslintrc.js`
- Replaces hardcoded browser targets with `BROWSER_TARGETS` constant across build configurations
- Updates `code/package.json` dependencies and browserslist configuration to reflect new targets



<!-- /greptile_comment -->

JReinhold committed:
update browser targets for Storybook 9.0

JReinhold committed:
Update Node to target v20 for Storybook 9

JReinhold committed:
add version compat linting

JReinhold committed:
update browser version data

JReinhold committed:
Merge branch 'next' into jeppe/browser-and-node-targets

ndelangen committed:
Merge branch 'next' into jeppe/browser-and-node-targets

ndelangen commented on code/core/src/builder-manager/index.ts:
```diff
@@ -28,6 +28,15 @@ const isRootPath = /^\/($|\?)/;
 let compilation: Compilation;
 let asyncIterator: ReturnType<StarterFunction> | ReturnType<BuilderFunction>;
 
+export const BROWSER_TARGETS = [
+  'chrome131',
+  'edge134',
+  'firefox136',
+  'safari18.3',
+  'ios18.3',
+  'opera117',
+];
```
This should be in the shared dir, maybe?

JReinhold replied:
Yeah I dont have a strong opinion on that.

ndelangen replied:
We need to (and i did it).

If the script to build needs this file, and that refers to dist files... that can't work.
So I moved these constants into a file without dependencies.

JReinhold replied:
Thank you ❤️

ndelangen commented on code/package.json:
```diff
@@ -165,6 +173,7 @@
     "esbuild-loader": "^4.3.0",
     "eslint": "8.57.1",
     "eslint-import-resolver-typescript": "^3.6.1",
+    "eslint-plugin-compat": "^6.0.2",
```
What does this do?

JReinhold replied:
It ensures that we're only using APIs that are compatible with the browsers defined in the browserslist configuration in code/package.json.

I'm thinking it's an experiment to see if it's useful or not.

ndelangen committed:
Refactor: Centralize browser and node target constants

- Moved BROWSER_TARGETS and NODE_TARGET constants to a new shared file for better organization and accessibility.
- Updated imports across various scripts to reference the new constants location.
- Removed redundant definitions from previous locations to streamline the codebase.

ndelangen committed:
Fix: Update imports for BROWSER_TARGETS and NODE_TARGET constants

- Adjusted import paths for BROWSER_TARGETS and NODE_TARGET in prep.ts and sourcefiles.ts to reflect their new location in the shared constants file.
- Ensured consistency in the codebase by removing outdated import statements.

ndelangen committed:
Refactor: Update import path for BROWSER_TARGETS in sourcefiles.ts

- Adjusted the import statement for BROWSER_TARGETS to reflect its new location in the shared constants file.
- Ensured consistency in the codebase by removing outdated import paths.
